Fix Flake: On TearDown script, delete all firewall-rules with a single gcloud run

azylinski opened the pull request:
#### What type of PR is this?

/kind bug
/kind flake

#### What this PR does / why we need it:

The kubemark-100 scalability tests fails on TearDown frequently (11% times - 13 out of 114 latest runs).
https://testgrid.k8s.io/sig-scalability-kubemark#kubemark-100

The fail happens as the script tries to delete subnetwork, which is still used by firewalls, see example buildlog:
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-kubemark-100-gce/1565151670835351552

In the code:
https://github.com/kubernetes/kubernetes/blob/5fa65e989bbc949c3cb4fb3a155d39bf49d8b9b1/cluster/gce/util.sh#L3758-L3760

So it looks like, the `delete-all-firewall-rules` is not fully finished.

The firewalls deletion happens in a loop and `kube::util::wait-for-jobs` guard in the end, which could be problematic
As a fix, I moved a delete all into a single gcloud command, which should be more solid.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

azylinski committed:
Fix: Refactor firewall-rules list to read-array

mborsz commented on cluster/gce/util.sh:
```diff
@@ -2576,9 +2576,11 @@ function create-cloud-nat-router() {
-    delete-firewall-rules "$fws"
+    # Deliberately do not quote, do not change unless a bug is found
+    # shellcheck disable=SC2086
+    gcloud compute firewall-rules delete --project "${NETWORK_PROJECT}" --quiet ${fws}
```
We don't need to make shellcheck unhappy, how about using pattern seen in this file, e.g. https://github.com/kubernetes/kubernetes/blob/1f07ce45325fc88da371b2b9b50e21623a5032a1/cluster/gce/util.sh#L3926-L3931

If you do something like
```
local -a fws
kube::util::read-array fws < <(gcloud compute firewall-rules list --project "${NETWORK_PROJECT}" --filter="network=${NETWORK}" --format="value(name)")
if (( "${#fws[@]}" > 0 )); then 
 echo "Deleting firewall rules remaining in network ${NETWORK}: ${fws}"
 delete-firewall-rules "${fws[@]}"
fi

this should probably work


```

azylinski replied:
Hi Maciej, I'm not sure, if I understand your comment.

The main point behind this PR is not to use the `delete-firewall-rules` func, which might not be as stable as it looks. Instead, I want to delete all firewalls at once with native `gcloud` command.
The linter issue is about `"${fws[@]}"` vs `${fws[@]}` - it has to be the 2nd option, as we're passing a inline, space separated list of firewalls names, e.g. `fw1 fw2 fw3 ...` - if you add quotation marks, it thinks, it's a single firewall name with spaces in a middle, which is not valid.

This pattern is used in many places, like flags:
https://github.com/kubernetes/kubernetes/blob/8a0e0c7f79d6d5c9b4acb0accbcb5bbee5cc9ef4/cluster/gce/gci/configure-helper.sh#L2380-L2381

Now, your suggestion is more around `kube::util::read-array` vs `format='value[terminator=" "](name)'`, which I don't have a strong opinion about and I'm happy to refactor, if you think it's cleaner.

azylinski replied:
Actually, looking closer at it, `kube::util::read-array` it's still an array, which you can iterate on (or e.g. check, it's size). I don't think it would work in this case - unless I don't see something?

mborsz replied:
> The main point behind this PR is not to use the delete-firewall-rules func, which might not be as stable as it looks. Instead, I want to delete all firewalls at once with native gcloud command.

OK, I think this is where we have misunderstanding.

IIUC the real bug here is that we call delete-firewall-rules with all firewalls in a single argument, which makes this call no-op, as there is no firewall with such name (there are no "Deleted [https://www.googleapis.com/compute/v1/projects/k8s-infra-e2e-boskos-scale-17/global/firewalls/XXX]" lines in the log which are printed by gcloud when called to delete object).

IMHO we should fix this bug. There is no need to migrate to custom gcloud call.

azylinski replied:
Ok, I see your point now.

All it all, I think it's all about calling `delete-firewall-rules`, which requires list of arguments (vs a single argument which is a list) - I think that was a main issue.

I'd go with your approach, as it requires less changes to the actual logic.

azylinski commented:
/restest

azylinski commented:
/retest

mborsz commented:
Looks OK now, can we test this somehow?

aojea commented:
/test

aojea commented:
/test pull-kubernetes-kubemark-e2e-gce-scale

aojea commented:
@mborsz is that job valid for testing this? ^^

mborsz commented:
I don't think we have any integration test that is good enough to trigger the issue it fixes.

I already discussed this with Artur and he will make some manual runs of specific code paths.

azylinski commented:
Ok, so I've tested it on my dev. cluster - try to manually run `delete-all-firewall-rules` with NETWORK env set.

Here're the outputs:

```
BEFORE, stdout:

Deleting firewall rules remaining in network default: gke-amce-7443d9aa-all
gke-amce-7443d9aa-ssh
gke-amce-7443d9aa-vms
Failed to delete firewall rules.

---

AFTER, stdout:

Deleting firewall rules remaining in network default: gke-amce-7443d9aa-all

Deleted [https://www.googleapis.com/compute/v1/projects/azylinski-gke-dev/global/firewalls/gke-amce-7443d9aa-all].
Deleted [https://www.googleapis.com/compute/v1/projects/azylinski-gke-dev/global/firewalls/gke-amce-7443d9aa-ssh].
Deleted [https://www.googleapis.com/compute/v1/projects/azylinski-gke-dev/global/firewalls/gke-amce-7443d9aa-vms].
```

That's a clear evidence, that our suspicion was correct and the param was passed incorrectly (before: multi-line list).

mborsz commented on cluster/gce/util.sh:
```diff
@@ -2576,11 +2576,13 @@ function create-cloud-nat-router() {
+  local -a fws
+  kube::util::read-array fws < <(gcloud compute firewall-rules list --project "${NETWORK_PROJECT}" --filter="network=${NETWORK}" --format="value(name)")
+  if (( "${#fws[@]}" > 0 )); then
     echo "Deleting firewall rules remaining in network ${NETWORK}: ${fws}"
```
This echo needs to be changed as well, to ${fws[@]} to print more than the first element.

azylinski replied:
Yes, I've noticed it at as. Updated.

azylinski replied:
Actually, that should be `${fws[*]}`
https://tldp.org/LDP/Bash-Beginners-Guide/html/Bash-Beginners-Guide.html#sect_10_02_02

mborsz commented:
/lgtm
/approve

mborsz commented:
Please fix the linter

mborsz commented:
/lgtm
/approve
/assign @mm4tt

mm4tt commented:
/approve
